docs/software/ml/pytorch.md

```diff
@@ -0,0 +1,392 @@
+
+!!! example "slurm sbatch script"
+    ```bash
+    #!/bin/bash
+    #SBATCH --job-name=myjob
+    #SBATCH --nodes=1
+    #SBATCH --ntasks-per-node=4
+    #SBATCH --cpus-per-task=72
+    #SBATCH --time=00:30:00
+    #SBATCH --uenv=pytorch/v2.6.0:/user-environment
+    #SBATCH --view=default
+
+    #################################
+    # OpenMP environment variables #
+    #################################
+    export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK  # (1)!
+
+    #################################
+    # PyTorch environment variables #
+    #################################
+    export MASTER_ADDR=$(hostname) # (2)!
+    export MASTER_PORT=6000
+    export WORLD_SIZE=$SLURM_NPROCS
+    export TORCH_NCCL_ASYNC_ERROR_HANDLING=1 # (3)!
+
+    #################################
+    # MPICH environment variables   #
+    #################################
+    export MPICH_GPU_SUPPORT_ENABLED=0 # (4)!
+
+    #################################
+    # CUDA environment variables    #
+    #################################
+    export CUDA_CACHE_DISABLE=1 # (5)!
+
+    ############################################
+    # NCCL and Fabric environment variables    #
+    ############################################
+    export NCCL_NET="AWS Libfabric" # (6)!
+    export NCCL_NET_GDR_LEVEL=PHB
+    export NCCL_CROSS_NIC=1
+    export FI_CXI_DISABLE_HOST_REGISTER=1
+    export FI_MR_CACHE_MONITOR=userfaultfd
+    export FI_CXI_DEFAULT_CQ_SIZE=131072
+    export FI_CXI_DEFAULT_TX_SIZE=32768
+    export FI_CXI_RX_MATCH_MODE=software
+
+    # (7)!
+    # (8)!
+    srun bash -c "
+        export RANK=\$SLURM_PROCID
+        export LOCAL_RANK=\$SLURM_LOCALID
+        . ./my-venv/bin/activate
+        python myscript.py
+    "
+    ```
+
+    1. Only set `OMP_NUM_THREADS` if you are using OpenMP in your code.
+    2. These variables are used by PyTorch to initialize the distributed
+       backend. The `MASTER_ADDR` and `MASTER_PORT` variables are used to
+       determine the address and port of the master node. Additionally we also need
+       `RANK` and `LOCAL_RANK` but these must be set per-process, see below.
+    3. Enable more graceful exception handling, see [PyTorch
+       documentation](https://pytorch.org/docs/stable/torch_nccl_environment_variables.html)
+    4. Disable GPU support in MPICH, as it [can lead to
+       deadlocks](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/mpi.html#inter-gpu-communication-with-cuda-aware-mpi)
+       when using together with nccl.
+    5. Avoid writing JITed binaries to the (distributed) file system, which
+       could lead to performance issues.
+    6. These variables should always be set for correctness and optimal
+       performance when using NCCL, see [the detailed
+       explanation][ref-communication-nccl].
+    7. `RANK` and `LOCAL_RANK` are set per-process by the SLURM job launcher.
+    8. Activate the virtual environment created on top of the uenv (if any).
```

Comment:
great and detailed script!
maybe we should also link to
https://docs.tds.cscs.ch/84/guides/storage/#python-virtual-environments-with-uenv (which also could be move elsewhere perhaps?)

```suggestion
    8. Activate the virtual environment created on top of the uenv (if any).

Please follow the guidelines for [python virtual environments with uenv][ref-guides-storage#python-virtual-environments-with-uenv] to enhance scalability and reduce load times.
```
not sure how to use hashes in urls here tho...

Reply:
good point - I'll try to fit it in

Reply:
Add an anchor to the subheading, and link to that directly. See https://eth-cscs.github.io/cscs-docs/contributing/#internal-links.